aiographite/aiographitesend.py

```diff
@@ -0,0 +1,449 @@
+SUPPORT_PROTOCOLS = ["plaintext", "pickle"]
+
+
+class AioGraphiteSendException(Exception):
+    pass
+
+
+class AsyncioGraphiteSendService(object):
+
+	def __init__(self, graphite_server, graphite_port, protocol = "pickle"):
+
+		self.graphite_server_address = (graphite_server, graphite_port)
+
+		self.connect_to_graphite()
+
+		self.protocol = protocol
+
+		self.loop = asyncio.get_event_loop()
+
+
+	def connect_to_graphite(self):
+		"""
+			Connect to Graphite Server based on Provided Server Address
+		"""
+		try:
+			self.socket = socket.create_connection(self.graphite_server_address)
+			self.socket.setblocking(False)
+		except socket.gaierror:
+			raise AioGraphiteSendException("Unable to connect to the provided server address %s:%s" % self.graphite_server_address)
+		except Exception as e:
+			raise e
+		return self.socket
+
+
+
+	def send_single_data(self, metric_dir_list, value, timestamp = None):
+		"""
+			@example: 
+				Assuming that 
+
+					Expected_Metric_Name  =  metaccounts.authentication.password.attempted
+
+				Then input metric_dir_list should be
+
+					metric_dir_list = [metaccounts, authentication, password, attempted]
+
+			@metric_dir_list: List of string
+			@timestamp: the type should be int
+
+			If you're very confident that the metric name is valid, then use <method: send_single_valid_data> instead.
+
+		"""		
+		valid_metric_name = self.to_graphite_valid_metric_name(metric_dir_list)
+		self.send_single_valid_data(valid_metric_name, value, timestamp)
+
+
+
+	def send_dataset_list(self, dataset, timestamp = None):
+		"""
+			@param: 
+				Support two kinds of dataset
+
+				1)	dataset = [(metric_dir_list, value1), (metric_dir_list, value2), ...] 
+
+				or 
+
+				2)	dataset = [(metric_dir_list1, value1, timestamp1), (metric_dir_list1, value2, timestamp2), ...]
+
+			If you're very confident that the metric name is valid, then use <method: send_valid_dataset_list> instead.
+
+		"""
+		if not dataset:
+			return 
+
+		if len(dataset[0]) == 2:
+			valid_dataset = [(self.to_graphite_valid_metric_name(metric_dir_list), value) for metric_dir_list, value in dataset]
+		else:
+			valid_dataset = [(self.to_graphite_valid_metric_name(metric_dir_list), value, timestamp) for metric_dir_list, value, timestamp in dataset]
+
+		self.send_valid_dataset_list(valid_dataset, timestamp)
+
+
+
+	def send_single_valid_data(self, metric, value, timestamp = None):
+		"""
+			@metric: String
+			@value: int
+			@timestamp: int
+			Send a single data(metric value timestamp) to graphite
+		"""
+		timestamp = int(time.time()) if timestamp is None else int(timestamp)
+		message = ""
+		if self.protocol == "plaintext":
+			message = plaintext_protocol_formatted_data(metric, value, timestamp)
+		else:
+			listOfMetricTuples = [pickle_protocol_formatted_data(metric, value, timestamp)]
+			message = generate_message_for_pickle(listOfMetricTuples)
+		self.loop.run_until_complete(asyncio.ensure_future(self.send_message(message)))
```

Comment:
It looks like sockets are synchronous, so they won't work really well here. 

Let's use asyncio's connections instead: that allows for asynchronous IO:
http://www.curiousefficiency.org/posts/2015/07/asyncio-tcp-echo-server.html